peps/pep-0733.rst

```diff
@@ -0,0 +1,660 @@
+[`Issue 62 <https://github.com/capi-workgroup/problems/issues/62>`__],
+or a new API tier of "blessed" functions
+[`Issue 55 <https://github.com/capi-workgroup/problems/issues/55>`__].
+
+
+API Specification and Abstraction
+---------------------------------
+
+The C API does not have a formal specification, it is described
+semi-formally in the documentation and exposed through C header
+files. This is also true for the limited API and the stable ABI.
+This creates a number of problems.
+
+Bindings for languages other than C/C++ must parse C code
+[`Issue 7 <https://github.com/capi-workgroup/problems/issues/7>`__].
+Some C language features are hard to handle in this way, because
+they produce compiler-dependent output (such as enums) or require
+a C preprocessor/compiler rather than just a parser (such as macros)
+[`Issue 35 <https://github.com/capi-workgroup/problems/issues/35>`__].
+
+Furthermore, C header files tend to expose more than what is intended
+to be part of the public API
+[`Issue 34 <https://github.com/capi-workgroup/problems/issues/34>`__].
+In particular, implementation details such as the precise memory
+layouts of internal data structures can be exposed
+[`Issue 22 <https://github.com/capi-workgroup/problems/issues/22>`__
+and :pep:`620`].
+This can make API evolution very difficult, in particular when it
+occurs in the stable ABI as in the case of ``ob_refcnt`` and ``ob_type``,
+which are accessed via the reference counting macros
+[`Issue 45 <https://github.com/capi-workgroup/problems/issues/45>`__].
+
+We identified a deeper issue in relation to the way that reference
+counting is exposed. The way that C extensions are required to
+manage references with calls to ``Py_INCREF`` and ``Py_DECREF`` is
+specific to CPython's memory model, and is hard for alternative
+Python implementations to emulate.
+[`Issue 12 <https://github.com/capi-workgroup/problems/issues/12>`__].
+
+Another set of problems arises from the fact that a ``PyObject*`` is
+exposed in the C API as an actual pointer rather than a handle. The
+address of an object serves as its ID and is used for comparison,
+and this complicates matters for alternative Python implementations
+that move objects during GC
+[`Issue 37 <https://github.com/capi-workgroup/problems/issues/37>`__].
+
+A separate issue is that object references are opaque to the runtime,
+discoverable only through calls to ``tp_traverse``/``tp_clear``,
+which have their own purposes. If there was a way for the runtime to
+know the structure of the object graph, and keep up with changes in it,
+this would make it possible for alternative implementations to implement
+different memory management schemes
+[`Issue 33 <https://github.com/capi-workgroup/problems/issues/33>`__].
+
+Object Reference Management
+---------------------------
+
+There are C API functions that return borrowed references, and
+functions that steal references to arguments, but there isn't a
```

Comment:
I still don't like the term steal. It is unnecessarily pejorative. 
"Stealing" a reference is just a transfer of ownership from caller to callee, which has many legitimate use cases,
particularly when the caller no longer needs the reference. "Stealing" reduces the number of decrefs and reduces the lifetime of unwanted objects.
Rust calls such a transfer of ownership a "move", but that is ownership of the object, not the reference.

If you really must use "steal", then please add a note explaining the term.

Reply:
Maybe for the purposes of this document, simply writing `"steal"` to acknowledge it's the commonly used term and doesn't necessarily have its usual implications is sufficient?

If we don't have a linkable glossary entry already, this isn't the place to write the first definition. (We probably have a linkable glossary entry though, so maybe quotes and a link to that will prevent people from being upset if this is where they first encounter the term.)

Reply:
I'll just say "decref arguments".